README.md
rodi_robot
==========
[RoDI] (http://rodibot.com/) (Robot Didáctico Inalámbrico) ROS package.

Allows to control a RoDI from the Robot Operating System (ROS) without the need
to flash a custom firmware and just using the default firmware web services API.

Author
------

Javier Martinez Canillas <javier@dowhile0.org>


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rodi_robot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.1 (2017-06-06)
------------------
* Initial release.


src/rodi_ws_node.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Range
from transport import Transport


def cmd_vel_cb(msg, transport):
    if msg.angular.z == 0 and msg.linear.x == 0:
        transport.stop()
        return

    if msg.linear.x > 0:
        transport.move_forward()
        return

    if msg.linear.x < 0:
        transport.move_reverse()
        return

    if msg.angular.z > 0:
        transport.move_left()
        return

    if msg.angular.z < 0:
        transport.move_right()
        return

if __name__ == '__main__':
    rospy.init_node('rodi_ws_node')

    transport = Transport()

    sub = rospy.Subscriber('cmd_vel', Twist, cmd_vel_cb, transport)
    pub = rospy.Publisher('ultrasound', Range, queue_size=1)

    rate = rospy.Rate(2)  # 2 Hz

    range = Range()
    range.radiation_type = 0  # ULTRASOUND
    range.header.frame_id = "/ultrasound"
    range.field_of_view = 0.52
    range.min_range = 0.2
    range.max_range = 4.0

    while not rospy.is_shutdown():
        try:
            range.range = float(transport.see()) / 100.0
            range.header.stamp = rospy.Time.now()
            pub.publish(range)
        except Exception as e:
            rospy.logerr("getting sonar data failed: " + str(e))

        rate.sleep()


src/rodi_node_node.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Range
from transport import Transport


def cmd_vel_cb(msg, transport):
    if msg.angular.z == 0 and msg.linear.x == 0:
        transport.stop()
        return

    if msg.linear.x > 0:
        transport.move_forward()
        return

    if msg.linear.x < 0:
        transport.move_reverse()
        return

    if msg.angular.z > 0:
        transport.move_left()
        return

    if msg.angular.z < 0:
        transport.move_right()
        return

if __name__ == '__main__':
    rospy.init_node('rodi_robot_node')

    transport = Transport()

    sub = rospy.Subscriber('cmd_vel', Twist, cmd_vel_cb, transport)
    pub = rospy.Publisher('ultrasound', Range, queue_size=1)

    rate = rospy.Rate(2)  # 2 Hz

    range = Range()
    range.radiation_type = 0  # ULTRASOUND
    range.header.frame_id = "/ultrasound"
    range.field_of_view = 0.52
    range.min_range = 0.2
    range.max_range = 4.0

    while not rospy.is_shutdown():
        try:
            range.range = float(transport.see()) / 100.0
            range.header.stamp = rospy.Time.now()
            pub.publish(range)
        except Exception as e:
            rospy.logerr("getting sonar data failed: " + str(e))

        rate.sleep()


src/rodi_robot/transport.py
#!/usr/bin/env python

from rospy import logdebug
from httplib import HTTPConnection


class Transport(object):

    def __init__(self, hostname='192.168.4.1', port='1234'):
        self.hostname = hostname
        self.port = port

    def send_command(self, params):
        request = "/" + "/".join(map(str, params))

        try:
            self.conn = HTTPConnection(self.hostname,
                                       port=self.port,
                                       timeout=0.5)
            self.conn.request("GET", request)
            response = self.conn.getresponse().read()
            self.conn.close()
            if len(response) == 0:
                return None
            return response
        except Exception as e:
            logdebug("the HTTP request " + request + " failed: " + str(e))
            return None

    def move_forward(self):
        self.send_command([3, 100, 100])

    def move_reverse(self):
        self.send_command([3, -100, -100])

    def move_left(self):
        self.send_command([3, -100, 100])

    def move_right(self):
        self.send_command([3, 100, -100])

    def stop(self):
        self.send_command([3, 0, 0])

    def see(self):
        return self.send_command([5])


src/rodi_robot/main.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Range
from rodi_robot.transport import Transport


class RodiRobot(object):

    ''' Robot class to add ROS support to RoDI'''

    def __init__(self):
        hostname = rospy.get_param('~hostname', '192.168.4.1')
        port = rospy.get_param('~port', '1234')
        self.timeout = rospy.get_param('~timeout', 2.0)

        self.transport = Transport(hostname, port)

        rospy.Subscriber('cmd_vel',
                         Twist,
                         self._cmd_vel_cb)
        self.pub = rospy.Publisher('ultrasound',
                                   Range,
                                   queue_size=1)

        self.rate = rospy.Rate(1)  # 1 Hz

        self.us_sensor = Range()
        self.us_sensor.radiation_type = 0  # ULTRASOUND
        self.us_sensor.header.frame_id = "/ultrasound"
        self.us_sensor.field_of_view = 0.52
        self.us_sensor.min_range = 0.2
        self.us_sensor.max_range = 1.0

        self.last_cmd_vel = rospy.Time.now()

    def _cmd_vel_cb(self, msg):
        self.last_cmd_vel = rospy.Time.now()
        if msg.angular.z == 0 and msg.linear.x == 0:
            self.transport.stop()
        elif msg.linear.x > 0:
            self.transport.move_forward()
        elif msg.linear.x < 0:
            self.transport.move_reverse()
        elif msg.angular.z > 0:
            self.transport.move_left()
        elif msg.angular.z < 0:
            self.transport.move_right()

    def run(self):
        while not rospy.is_shutdown():
            if (rospy.Time.now() - self.last_cmd_vel).secs >= self.timeout:
                self.transport.stop()
                self.last_cmd_vel = rospy.Time.now()
            try:
                range = self.transport.see()
                if range is not None:
                    self.us_sensor.range = float(range) / 100.0
                    self.us_sensor.header.stamp = rospy.Time.now()
                    self.pub.publish(self.us_sensor)
            except Exception as exception:
                rospy.logerr("getting sonar data failed: " + str(exception))

            self.rate.sleep()
